src/core/Ekubo.sol

```diff
@@ -0,0 +1,372 @@
+        }
+        if (feeOnTransfer) {
+            state.globalSell.amount =
+                _ekuboPay(state.globalSell.token, payer, state.globalSell.amount, permit, isForwarded, sig);
+        }
+        if (state.globalSell.amount == 0) {
+            revert ZeroSellAmount(state.globalSell.token);
+        }
+        state.globalSellAmount = state.globalSell.amount;
+        data = newData;
+
+        PoolKey memory poolKey;
+
+        while (data.length >= _HOP_DATA_LENGTH) {
+            // TODO: Check if BPS is needed
+            uint16 bps;
+            assembly ("memory-safe") {
+                bps := shr(0xf0, calldataload(data.offset))
+
+                data.offset := add(0x02, data.offset)
+                data.length := sub(data.length, 0x02)
+                // we don't check for array out-of-bounds here; we will check it later in `Decoder.overflowCheck`
+            }
+
+            data = Decoder.updateState(state, notes, data);
+            // TODO: check the sign convention
+            int256 amountSpecified = int256((state.sell.amount * bps).unsafeDiv(BASIS)).unsafeNeg();
+            bool zeroForOne;
+            {
+                (IERC20 sellToken, IERC20 buyToken) = (state.sell.token, state.buy.token);
+                assembly ("memory-safe") {
+                    sellToken := and(_ADDRESS_MASK, sellToken)
+                    buyToken := and(_ADDRESS_MASK, buyToken)
+                    zeroForOne :=
+                        or(
+                            eq(sellToken, 0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee),
+                            and(iszero(eq(buyToken, 0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee)), lt(sellToken, buyToken))
+                        )
+                }
+                (poolKey.token0, poolKey.token1) = zeroForOne.maybeSwap(address(sellToken), address(buyToken));
+            }
+
+            {
+                bytes32 config;
+                assembly ("memory-safe") {
+                    config := calldataload(data.offset)
+                    data.offset := add(0x20, data.offset)
+                    data.length := sub(data.length, 0x20)
+                    // we don't check for array out-of-bounds here; we will check it later in `Decoder.overflowCheck`
+                }
+                poolKey.config = Config.wrap(config);
+            }
+
+            uint256 skipAhead;
+            assembly ("memory-safe") {
+                skipAhead := calldataload(data.offset)
+                data.offset := add(0x20, data.offset)
+                data.length := sub(data.length, 0x20)
+                // we don't check for array out-of-bounds here; we will check it later in `Decoder.overflowCheck`
+            }
+
+            Decoder.overflowCheck(data);
+
+            {
+                SqrtRatio sqrtRatio =
+                    SqrtRatio.wrap(uint96(zeroForOne.ternary(uint256(4611797791050542631), uint256(79227682466138141934206691491))));
```

Comment:
:eyes: funky constants. I assume that these are different due to Ekubo's weird fixnum encoding?

Reply:
Yep, taken from their contract
<img width="737" alt="image" src="https://github.com/user-attachments/assets/3d447cb9-a840-4d2b-bd88-66685a16c08e" />